R/get_vimc_climate.R

```diff
@@ -0,0 +1,150 @@
+#' character code. E.g. one of "Canada", "CA", or "CAN". Input is checked
+#' against names and codes in \pkg{countrycode}.
+#' @param date_range A two element vector of either strings or `<Date>`s, giving
+#' the interval for which to get data. If a character vector, must be in a
+#' format that can be converted to a `<Date>` by [lubridate::as_date()].
+#' @param data_location A location relative to the current working directory
+#' to search for the data. Data for `country` are expected to live in
+#' `data_location/XYZ` as Parquet files, where `XYZ` is the ISO 3-character code
+#' for `country`. Defaults to the working directory.
+#' @param data_source A string giving the climate data to search for. Names must
+#' match exactly.
+#' @param admin_level A numeric giving the GADM admin level for which to get
+#' data. Only one unit may be passed at a time.
+#'
+#' @return A `<data.frame>` of climate data.
+#'
+#' @export
+get_vimc_climate <- function(
+  country,
+  date_range,
+  data_location = here::here(),
+  data_source = c(
+    "CHIRPS",
+    "ERA5_mean",
+    "ERA5_min",
+    "ERA5_max",
+    "ERA5_RH",
+    "ERA5_SH",
+    "PERSIANN"
+  ),
+  admin_level = c(0, 1, 2, 3)
+) {
+  # more checks below
+  checkmate::assert_string(
+    country,
+    min.chars = 2
+  )
+
+  # allow date as string because actually using Date class is annoying
+  if (!checkmate::test_vector(date_range, len = 2)) {
+    cli::cli_abort(
+      "`date_range` must be a 2-element vector; currently has a length of
+      length(date_range)"
+    )
+  }
+  is_good_daterange <- checkmate::test_date(
+    date_range,
+    any.missing = FALSE
+  ) ||
+    checkmate::test_character(
+      date_range
+    )
+  if (!is_good_daterange) {
+    cli::cli_abort(
+      "`date_range` must be either a `<Date>` or `string` vector, but is a
+      {.cls {class(date_range)}}"
+    )
+  } else if (is.character(date_range)) {
+    # convert to Date
+    date_range <- lubridate::as_date(date_range)
+  }
+
+  # NOTE: need to check for sensible date values and throw an error
+  # if data are not available in the range
+
+  # allow only one choice at a time to avoid ambiguous return type
+  data_source <- rlang::arg_match(data_source)
+  admin_level <- as.character(admin_level)
+  admin_level <- rlang::arg_match(admin_level, allowed_admin_levels)
+
+  # convert country to ISO 3166-1 3C country code
+  # note `nomatch = NULL` is preferable but is a special case
+  # in pkg countrycode
+  nchar_country <- as.character(nchar(country))
+  country_iso3c <- switch(
+    nchar_country,
+    "2" = countrycode::countrycode(
+      country,
+      "iso2c",
+      "iso3c",
+      warn = FALSE,
+      nomatch = NA
+    ),
+    "3" = countrycode::countrycode(
+      country,
+      "iso3c",
+      "iso3c",
+      warn = FALSE,
+      nomatch = NA
+    ),
+    countrycode::countryname(
+      country,
+      "iso3c",
+      warn = FALSE,
+      nomatch = NA
+    )
+  )
```

Comment:
Move all this into a validate_country_code function or similar.  Please use an if/else ladder rather than switch with a conversion from integer to character, the semantics are not obvious